pkg/processor/procbuiltin/unwrap.go

```diff
@@ -0,0 +1,350 @@
+	// return the original key if it doesn't contain a payload
+	if !ok {
+		return key
+	}
+
+	// if payload is a map, return the payload as structured data
+	if p, ok := payload.(map[string]any); ok {
+		return record.StructuredData(p)
+	}
+
+	// otherwise, convert the payload to string, then return it as raw data
+	raw := fmt.Sprint(payload)
+
+	return record.RawData{Raw: []byte(raw)}
+}
+
+/*
+	 example of debezium record:
+		`{
+		 "payload": {
+		   "after": {
+		     "description": "desc",
+		     "id": 20
+		   },
+		   "before": null,
+		   "op": "c",
+		   "source": {
+		     "opencdc.readAt": "1674061777225877000",
+		     "opencdc.version": "v1",
+		   },
+		   "transaction": null,
+		   "ts_ms": 1674061777225
+		 },
+		 "schema": {} // will be ignored
+		}`
+*/
+// DebeziumUnwrapper unwraps a debezium record from the payload, expects rec.Payload.After to be of type record.StructuredData
+type DebeziumUnwrapper struct{}
+
+const (
+	debeziumOpCreate = "c"
+	debeziumOpUpdate = "u"
+	debeziumOpDelete = "d"
+	debeziumOpRead   = "r" // snapshot
+
+	debeziumFieldBefore    = "before"
+	debeziumFieldAfter     = "after"
+	debeziumFieldSource    = "source"
+	debeziumFieldOp        = "op"
+	debeziumFieldTimestamp = "ts_ms"
+)
+
+func (d *DebeziumUnwrapper) Unwrap(rec record.Record) (record.Record, error) {
+	// record must be structured
+	structPayload, ok := rec.Payload.After.(record.StructuredData)
+	if !ok {
+		return record.Record{}, cerrors.Errorf("record payload data must be structured data")
+	}
+	// get payload
+	payload := structPayload["payload"] // the payload has the debezium record
+	if p, ok := payload.(map[string]interface{}); ok {
+		structPayload = p
+	} else {
+		return record.Record{}, cerrors.Errorf("payload doesn't contain a record")
+	}
+
+	// check fields under payload
+	err := d.validateRecord(structPayload)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	var before record.StructuredData
+	beforeData := structPayload[debeziumFieldBefore]
+	b, ok := beforeData.(map[string]any)
+	if beforeData != nil && !ok {
+		return record.Record{}, cerrors.Errorf("%s field is not a map", debeziumFieldBefore)
+	}
+	before = b
+
+	var after record.StructuredData
+	afterData := structPayload[debeziumFieldAfter]
+	a, ok := afterData.(map[string]any)
+	if afterData != nil && !ok {
+		return record.Record{}, cerrors.Errorf("%s field is not a map", debeziumFieldAfter)
+	}
+	after = a
+
+	op, ok := structPayload[debeziumFieldOp].(string)
+	if !ok {
+		return record.Record{}, cerrors.Errorf("%s operation is not a string", op)
+	}
+
+	operation, err := d.convertOperation(op)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping operation: %w", err)
+	}
+
+	metadata, err := d.unwrapMetadata(rec)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping metadata: %w", err)
+	}
+
+	// create a kafka connect unwrapper to unwrap the key
+	kafkaConnectUnwrapper := &KafkaConnectUnwrapper{}
```

Comment:
Nitpick: To prevent us creating a new struct for each record we could have a field of type `KafkaConnectUnwrapper` in `DebeziumUnwrapper` and use that like `d.kafkaConnectUnwrapper.UnwrapKey(key)`.